File: java/clock_tree/clock_tree_icc.java
import java.util.* ;
import java.util.regex.Matcher ;
import java.util.regex.Pattern ;
import java.io.* ;
//import CTSNode ;


public class clock_tree_icc {

    // class
    public static Boolean debug = false ;

    public static CTSNode read_cts_icc ( String filename ) throws Exception
    {
        FileReader readfile = new FileReader( filename ) ;
        BufferedReader readbuf = new BufferedReader( readfile ) ;

        CTSNode CTS = new CTSNode( filename , null ) ;
        String line = "" ;
        String clk = "" ;
        Matcher m ;
        Pattern reRoot = Pattern.compile( "Printing structure within exceptions of (.*) at root pin" ) ;
        Pattern reTreeStructure = Pattern.compile( " *.([0-9]+). (.*?)" ) ;
        ArrayList<CTSNode> hier = new ArrayList<>() ;

        while ( (line=readbuf.readLine()) != null )
        {
            // clock declaration
            m = reRoot.matcher(line) ;
            if (m.find())
            {
                clk = m.group(1) ;
                CTSNode CLK = new CTSNode( clk , CTS ) ;
                hier.clear() ;
                hier.add( CLK ) ;
                continue ;
            }

            if (clk=="") continue ;

            // line matches tree structure
            m = reTreeStructure.matcher(line) ;
            if (! m.find()) continue ;

            int level = Integer.parseInt(m.group(1)) ;
            String inst = m.group(2) ;

            // test level and hier
            if (level <= (hier.size()-1))  // step up in hier
                for (int i=hier.size()-1 ;  i>level ; i--)
                    hier.remove(i);
            CTSNode UP = hier.get( hier.size()-1 ) ;
            CTSNode CURRENT = new CTSNode( line , UP ) ;
            hier.add( CURRENT ) ;

            if (debug)
                System.out.println( "\nUP = "+UP+"\nCURRENT = "+CURRENT );
        }

        return CTS ;
    }

    public static void main( String args[] ) throws Exception
    {
        debug = true ;
        CTSNode.debug = true ;

        CTSNode TOP = read_cts_icc( "../../py/cts/clk_main.txt" ) ;
        TOP.print_tree() ;
    }

}

File: py/clock_tree_browser/clock_tree_icc.py

import re
from clock_tree_CTSNode import CTSNode

debug = False


########################################
## read file
## return pointer to top-most CTSNode
##
## note : use upper case for CTSNode's

def read_cts_icc( filename , TOP=None ):

    global debug

    print( "Info : read_cts_icc '%s'" % filename )

    F = open(filename,'r')
    if TOP == None :
        CTS = CTSNode( filename, None, level=-2 )
    else :
        CTS = CTSNode( filename, TOP )
    clk = ''

    for line in F :

        ## clock declaration
        m = re.match('Printing structure within exceptions of (\w+) at root pin ',line)
        if m :
            clk = m.group(1)
            CLK = CTSNode( clk, CTS )
            hier = list()
            hier.append(CLK)
            continue

        if clk == '' : continue

        ## tool warnings
        if re.match('.* reconvergent clock path found',line) :
            print("Warning : " + clk + line)
            continue

        ## line matches tree structure
        m = re.match(' *\((\d+)\) (.*?)( |\[\w+:)',line)
        if not m : continue

        level = m.group(1)
        inst = m.group(2)

        ## test level + hier
        level = int(level)
        if level+1 < len(hier) : # step up in hier
            hier = hier[0:level+1]  # level included

        UP = hier[-1]
        CURRENT = CTSNode( line.rstrip() , UP )
        hier.append(CURRENT)

        if debug :
            print "\nUP = %s\nCURRENT = %s" % (UP,CURRENT)

    F.close()

    CTS.statistics()

    return CTS



########################################
## standalone __main__

if ( __name__ == "__main__" ) :

    debug = False

    TOP = read_cts_icc( 'clock_tree_report_icc_init_design_CTS_scenario/clk_ahb.txt' )

    print "\n==== full tree ===="
    TOP.print_tree()

    print "\n==== level 2 ===="
    TOP.find_tree( hideall=True , level=2 )
    TOP.print_tree()

    print "\n==== level 2 + NVM ===="
    TOP.find_tree( hideall=False , name='.*i_nvm_block/clk' )
    TOP.print_tree()


